lib/rucio/core/transfer.py

```diff
@@ -1125,19 +1125,19 @@ def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[
 @transactional_session
 def build_transfer_paths(
         topology: "Topology",
-        protocol_factory: "ProtocolFactory",
+        protocol_factory: ProtocolFactory,
         requests_with_sources: "Iterable[RequestWithSources]",
-        admin_accounts: "Optional[set[InternalAccount]]" = None,
-        schemes: "Optional[list[str]]" = None,
-        failover_schemes: "Optional[list[str]]" = None,
+        admin_accounts: Optional[set[InternalAccount]] = None,
+        schemes: Optional[list[str]] = None,
+        failover_schemes: Optional[list[str]] = None,
         max_sources: int = 4,
-        transfertools: "Optional[list[str]]" = None,
+        transfertools: Optional[list[str]] = None,
         requested_source_only: bool = False,
         preparer_mode: bool = False,
         *,
         session: "Session",
-        logger: "Callable" = logging.log,
-):
+        logger: LoggerFunction = logging.log,
+) -> "tuple[dict[str, Sequence[DirectTransfer]], set[str], set[str], set[str], set[str]]":
```

Comment:
All of the `str`s in this return type are `rws.request_id`, so to be compliant with its type this should be:

```suggestion
) -> "tuple[dict[Optional[str], Sequence[DirectTransfer]], set[Optional[str]], set[Optional[str]], set[Optional[str]], set[Optional[str]]]":
```

To confirm this, on line 1029, where all the returned elements are initialized:

```python
candidate_paths_by_request_id, reqs_no_source, reqs_only_tape_source, reqs_scheme_mismatch = {}, set(), set(), set()
```

If you were to annotate their type at initialization time, e.g.:

```python
reqs_no_source: set[str] = set()
```

You'd get this error when a `request_id` is added to them (e.g., when `reqs_no_source.add(rws.request_id)` happens):

```
Type "str | None" is not assignable to type "str"
```

However, I think your return type annotation is correct, and we should keep it like this. The 'compliant' type annotation wouldn't make much sense - `dict` expects unique key and `set` expects unique elements, so what happens if there's multiple `request_id`s that are `None`?

I think there's a pre-existing issue that we need to fix somewhere else. My impression is that `RequestWithSources` should always be expected to have a `request_id`, so the best fix would likely be to change that type from `Optional[str]` to `str`. 
If my impression is wrong and `RequestWithSources` can have a missing `request_id`, then we'd need to modify this function (`build_transfer_paths`) to check that `rws.request_id is not None` before adding it to the returned dict/sets (and we would probably want to log a message in case `rws.request_id is None`).